### PR TITLE
#330, #353: improved type resolution heuristics and added tests

### LINKED ISSUE #330: type inference weakness: fails to unify inferred template params across case statement
In this code, one would expect that the use of `None` would be inferred to have type `Option<int>` and therefore `T(none)` would have type `T<int>`, thus matching the return type and also the type from the other case.
```
datatype Option<A> = Some(value: A) | None

datatype T<A> = T(value: Option<A>)

function f(t: T<int>) : T<int>
{
  match t.value {
    case Some(val) => T(Some(val))
    case None => T(None)
  }
}
```
Unfortunately we get this error message:
```
$ dafny a.dfy 
Dafny 2.3.0.10506
a.dfy(9,18): Error: type of case bodies do not agree (found T<?>, previous types T<int>) (invariant type parameter would require int = ?)
1 resolution/type errors detected in a.dfy
```

Well, this wouldn't be so bad if we could specify the type parameter
either with `T(None<int>)` or `T<int>(None)`, but neither of these work and just give an error that these constructors don't take type params. This was the easiest workaround I could find:

```
datatype Option<A> = Some(value: A) | None

datatype T<A> = T(value: Option<A>)

function f(t: T<int>) : T<int>
{
  match t.value {
    case Some(val) => T(Some(val))
    case None => (
      var none : Option<int> := None;
      T(none)
    )
  }
}
```

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #330. Verified fix and added tests

### CODE REVIEW

File: Test/git-issues/git-issue-330.dfy [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+datatype T<A> = T(value: Option<A>)
+
+function f(t: T<int>) : T<int>
+{
+  match t.value {
+    case Some(val) => T(Some(val))
+    case None => (
+      var none : Option<int> := None;
+      T(none)
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-330.dfy:15]: The bug report listed this as the workaround, not the bug. The bug was that it was not possible to write just `T(None)` for L14-15. It seems that bug is still present.

**Comment by davidcok** [comment]
Fixes #330 #353.

File: Test/git-issues/git-issue-330.dfy (hunk 1 above)
Comment (davidcok) [reply Test/git-issues/git-issue-330.dfy:23]: Fixed the test, and included the workarounds as additional tests

**Comment by RustanLeino** [comment]
The test suite fails on `Test/dafny0/TypeTests.dfy`, where the resolver apparently loops forever. The relevant lines in the test that's causing the looping are:

```
  datatype Dt<X> = Ctor(X -> Dt<X>)
  method M1()
  {
    var dt: Dt;
    var f := x => x;
    dt := Ctor(f);  // error: cannot infer a type for f
  }
```

(The comment in this snippet is in the test file itself.)

**Comment by davidcok** [comment]
Fixed the non-terminating loop

**Pull request merged**